src/fspec-fromdir.c

```diff
@@ -1,85 +1,111 @@
-    printf("type=%s\n", filetype(sb->st_mode));
-    printf("mode=%04o\n", sb->st_mode & ~S_IFMT);
+static void printentry(char *, size_t, size_t);
 
-    if (!root_owned) {
-        if (sb->st_uid != 0)
-            printf("uid=%d\n", sb->st_uid);
-        if (sb->st_gid != 0)
-            printf("gid=%d\n", sb->st_gid);
+static void
+recurse(char *path, size_t len, size_t max)
+{
+    struct dirent **d;
+    int dlen;
+
+    dlen = scandir(path, &d, skipdot, alphasort);
+    if (dlen < 0)
+        err(1, "scandir %s", path);
+    if (len == max)
+        errx(1, "path is too long");
+    path[len++] = '/';
+    for (int i = 0; i < dlen; ++i) {
+        char *end = memccpy(path + len, d[i]->d_name, '\0', max - len);
+        if (!end)
+            errx(1, "path is too long");
+        printentry(path, end - path - 1, max);
+        free(d[i]);
     }
+    free(d);
+}
 
-    if (S_ISLNK(sb->st_mode)) {
-        len = readlink(fpath, pathbuf, sizeof(pathbuf));
-        if (len < 0)
-            err(1, "readlink of %s failed", fpath);
-        if (len == sizeof(pathbuf))
-            errx(1, "link target of %s too long", fpath);
-        pathbuf[len] = 0;
-        if (strchr(pathbuf, '\n'))
+static void
+printentry(char *path, size_t len, size_t max)
+{
+    static char buf[PATH_MAX];
+    ssize_t buflen;
+    struct stat st;
+
+    if (stat(path, &st) != 0)
+        err(1, "stat %s", path);
+
+    printf("%s%s\n", prefix, path + 2);
+
+    switch (st.st_mode & S_IFMT) {
+    case S_IFREG:
+        puts("type=reg");
+        if (absolute) {
+            if (realpath(path, buf) == NULL)
+                err(1, "realpath of %s failed", path);
+            if (strchr(buf, '\n'))
+                errx(1, "file path contains new line, aborting");
+            printf("source=%s\n", buf);
+        }
+        break;
+    case S_IFLNK:
+        puts("type=sym");
+        buflen = readlink(path, buf, sizeof(buf));
+        if (buflen < 0)
+            err(1, "readlink of %s failed", path);
+        if (buflen >= sizeof(buf))
+            errx(1, "link target of %s too long", path);
+        buf[buflen] = 0;
+        if (strchr(buf, '\n'))
             errx(1, "link target contains new line");
-        printf("link=%s\n", pathbuf);
+        printf("link=%s\n", buf);
+        break;
+    case S_IFCHR:
+    case S_IFBLK:
+        printf("type=%s\n", S_ISCHR(st.st_mode) ? "chardev" : "blockdev");
+        printf("devnum=%llu\n", (long long unsigned)st.st_rdev);
+        break;
+    case S_IFDIR:
+        puts("type=dir");
+        break;
     }
 
-    if (S_ISREG(sb->st_mode) && absolute) {
-        if (realpath(fpath, pathbuf) == NULL)
-            err(1, "realpath of %s failed", fpath);
-        if (strchr(pathbuf, '\n'))
-            errx(1, "file path contains new line, aborting");
-        printf("source=%s\n", pathbuf);
+    printf("mode=%04o\n", st.st_mode & ~S_IFMT);
+    if (!root_owned) {
+        if (st.st_uid != 0)
```

Comment:
Following our other convention, this printing should now be unconditional.

Reply:
I was thinking the same thing. Done.